### PR TITLE
Adding monthly and yearly arguments to `guess_bounds`

### LINKED ISSUE #4864: Should `guess_bounds` "do what I mean" for Gregorian monthly data?
## ✨ Feature Request
Iris coordinate `guess_bounds` functionality identifies the bounds as halfway between the points.  For a time coordinate, when the points are mid-month with a Gregorian calendar, should iris instead set the bounds to start and end of the month?

## Motivation
In ANTS, we have this functionality already, but do not support any other time cases.  Iris' `guess_bounds` is more flexible in time handling, so we'd love to retire our limited time handling in favour of using the iris behaviour.   In other words, iris handles the general case well, but does not handle this specific case as a user might expect; while ANTS handles this particular case well, but does not handle any other cases for time coordinates.  Ideally, iris `guess_bounds` would give us the best of both worlds.

We can delegate to iris `guess_bounds` to get the best of both worlds in ANTS.  I think the optimal solution though is for this behaviour to be available for all iris users.

ANTS docs are here, for reference: https://code.metoffice.gov.uk/doc/ancil/ants/latest/lib/ants.utils.html#ants.utils.coord.guess_bounds (and link through to the implementation source code - there's also a unit test for the Gregorian case here:https://code.metoffice.gov.uk/trac/ancil/browser/ants/trunk/lib/ants/tests/utils/coord/test_guess_bounds.py?marks=100-111#L100 )

This comes up fairly frequently when working with model data.

## Additional context
Example of current behaviour with iris 3.2: 
```
In [1]: import iris

In [2]: import iris.coords

In [3]: time = iris.coords.DimCoord(points=[210756., 211464., 212172., 212904., 213636., 214368., 215100
   ...: .,215844., 216576., 217308., 218040., 218772.], units='hours since epoch', standard_name='time')
   ...: 

In [4]: time.units.num2date(time.points)  # Points are mid-month for gregorian calendar
Out[4]: 
array([cftime.DatetimeGregorian(1994, 1, 16, 12, 0, 0, 0, has_year_zero=False),
       cftime.DatetimeGregorian(1994, 2, 15, 0, 0, 0, 0, has_year_zero=False),
       cftime.DatetimeGregorian(1994, 3, 16, 12, 0, 0, 0, has_year_zero=False),
       cftime.DatetimeGregorian(1994, 4, 16, 0, 0, 0, 0, has_year_zero=False),
       cftime.DatetimeGregorian(1994, 5, 16, 12, 0, 0, 0, has_year_zero=False),
       cftime.DatetimeGregorian(1994, 6, 16, 0, 0, 0, 0, has_year_zero=False),
       cftime.DatetimeGregorian(1994, 7, 16, 12, 0, 0, 0, has_year_zero=False),
       cftime.DatetimeGregorian(1994, 8, 16, 12, 0, 0, 0, has_year_zero=False),
       cftime.DatetimeGregorian(1994, 9, 16, 0, 0, 0, 0, has_year_zero=False),
       cftime.DatetimeGregorian(1994, 10, 16, 12, 0, 0, 0, has_year_zero=False),
       cftime.DatetimeGregorian(1994, 11, 16, 0, 0, 0, 0, has_year_zero=False),
       cftime.DatetimeGregorian(1994, 12, 16, 12, 0, 0, 0, has_year_zero=False)],
      dtype=object)

In [5]: time.guess_bounds()

In [6]: time.units.num2date(time.bounds)  # Bounds are not start/end of month
Out[6]: 
array([[cftime.DatetimeGregorian(1994, 1, 1, 18, 0, 0, 0, has_year_zero=False),
        cftime.DatetimeGregorian(1994, 1, 31, 6, 0, 0, 0, has_year_zero=False)],
       [cftime.DatetimeGregorian(1994, 1, 31, 6, 0, 0, 0, has_year_zero=False),
        cftime.DatetimeGregorian(1994, 3, 1, 18, 0, 0, 0, has_year_zero=False)],
       [cftime.DatetimeGregorian(1994, 3, 1, 18, 0, 0, 0, has_year_zero=False),
        cftime.DatetimeGregorian(1994, 3, 31, 18, 0, 0, 0, has_year_zero=False)],
       [cftime.DatetimeGregorian(1994, 3, 31, 18, 0, 0, 0, has_year_zero=False),
        cftime.DatetimeGregorian(1994, 5, 1, 6, 0, 0, 0, has_year_zero=False)],
       [cftime.DatetimeGregorian(1994, 5, 1, 6, 0, 0, 0, has_year_zero=False),
        cftime.DatetimeGregorian(1994, 5, 31, 18, 0, 0, 0, has_year_zero=False)],
       [cftime.DatetimeGregorian(1994, 5, 31, 18, 0, 0, 0, has_year_zero=False),
        cftime.DatetimeGregorian(1994, 7, 1, 6, 0, 0, 0, has_year_zero=False)],
       [cftime.DatetimeGregorian(1994, 7, 1, 6, 0, 0, 0, has_year_zero=False),
        cftime.DatetimeGregorian(1994, 8, 1, 0, 0, 0, 0, has_year_zero=False)],
       [cftime.DatetimeGregorian(1994, 8, 1, 0, 0, 0, 0, has_year_zero=False),
        cftime.DatetimeGregorian(1994, 8, 31, 18, 0, 0, 0, has_year_zero=False)],
       [cftime.DatetimeGregorian(1994, 8, 31, 18, 0, 0, 0, has_year_zero=False),
        cftime.DatetimeGregorian(1994, 10, 1, 6, 0, 0, 0, has_year_zero=False)],
       [cftime.DatetimeGregorian(1994, 10, 1, 6, 0, 0, 0, has_year_zero=False),
        cftime.DatetimeGregorian(1994, 10, 31, 18, 0, 0, 0, has_year_zero=False)],
       [cftime.DatetimeGregorian(1994, 10, 31, 18, 0, 0, 0, has_year_zero=False),
        cftime.DatetimeGregorian(1994, 12, 1, 6, 0, 0, 0, has_year_zero=False)],
       [cftime.DatetimeGregorian(1994, 12, 1, 6, 0, 0, 0, has_year_zero=False),
        cftime.DatetimeGregorian(1994, 12, 31, 18, 0, 0, 0, has_year_zero=False)]],
      dtype=object)
```

and similar for ANTS 0.19:

```
In [1]: import iris

In [2]: import ants

In [3]: time = iris.coords.DimCoord(points=[210756., 211464., 212172., 212904., 213636., 214368., 215100
   ...: .,215844., 216576., 217308., 218040., 218772.], units='hours since epoch', standard_name='time')
   ...: 

In [4]: time.units.num2date(time.points)  # Points are mid-month for gregorian calendar
Out[4]: 
array([real_datetime(1994, 1, 16, 12, 0),
       real_datetime(1994, 2, 15, 0, 0),
       real_datetime(1994, 3, 16, 12, 0),
       real_datetime(1994, 4, 16, 0, 0),
       real_datetime(1994, 5, 16, 12, 0),
       real_datetime(1994, 6, 16, 0, 0),
       real_datetime(1994, 7, 16, 12, 0),
       real_datetime(1994, 8, 16, 12, 0),
       real_datetime(1994, 9, 16, 0, 0),
       real_datetime(1994, 10, 16, 12, 0),
       real_datetime(1994, 11, 16, 0, 0),
       real_datetime(1994, 12, 16, 12, 0)], dtype=object)

In [5]: ants.utils.coord.guess_bounds(time)

In [6]: time.units.num2date(time.bounds)  # Bounds are now start/end of month
Out[6]: 
array([[real_datetime(1994, 1, 1, 0, 0), real_datetime(1994, 2, 1, 0, 0)],
       [real_datetime(1994, 2, 1, 0, 0), real_datetime(1994, 3, 1, 0, 0)],
       [real_datetime(1994, 3, 1, 0, 0), real_datetime(1994, 4, 1, 0, 0)],
       [real_datetime(1994, 4, 1, 0, 0), real_datetime(1994, 5, 1, 0, 0)],
       [real_datetime(1994, 5, 1, 0, 0), real_datetime(1994, 6, 1, 0, 0)],
       [real_datetime(1994, 6, 1, 0, 0), real_datetime(1994, 7, 1, 0, 0)],
       [real_datetime(1994, 7, 1, 0, 0), real_datetime(1994, 8, 1, 0, 0)],
       [real_datetime(1994, 8, 1, 0, 0), real_datetime(1994, 9, 1, 0, 0)],
       [real_datetime(1994, 9, 1, 0, 0),
        real_datetime(1994, 10, 1, 0, 0)],
       [real_datetime(1994, 10, 1, 0, 0),
        real_datetime(1994, 11, 1, 0, 0)],
       [real_datetime(1994, 11, 1, 0, 0),
        real_datetime(1994, 12, 1, 0, 0)],
       [real_datetime(1994, 12, 1, 0, 0),
        real_datetime(1995, 1, 1, 0, 0)]], dtype=object)
```

### PR DESCRIPTION
**Pull request by HGWright**
### Description

---

Closes #4864

This adds the `monthly` and `yearly` options to `guess_bounds`.

---
Add any of the below labels to trigger actions on this PR:

- https://github.com/SciTools/iris/labels/benchmark_this

### CODE REVIEW

File: lib/iris/coords.py [hunk 1]
```diff
@@ -2233,31 +2233,63 @@ def _guess_bounds(self, bound_position=0.5):
...
+                if date.month == 12:
+                    lyear = date.year
+                    uyear = date.year + 1
+                    lmonth = 12
+                    umonth = 1
+                else:
+                    lyear = uyear = date.year
+                    lmonth = date.month
+                    umonth = date.month + 1
+                # if monthly:
```
Comment (pp-mo) [inline lib/iris/coords.py:2251]: Obsolete comment. I guess

**Review by pp-mo — CHANGES_REQUESTED** [review]
Monthly looks good.
I checked the coverage + it is good.

Do you want to try for yearly too ?!?

File: lib/iris/coords.py [hunk 2]
```diff
@@ -2233,31 +2233,63 @@ def _guess_bounds(self, bound_position=0.5):
...
+                if f"{date.month} + {date.year}" not in months_and_years:
+                    months_and_years.append(f"{date.month} + {date.year}")
```
[the remark this answers is not in the document]
Comment (HGWright) [reply lib/iris/coords.py]: [0f97464](https://github.com/SciTools/iris/pull/6090/commits/0f97464eea07abf00bf7a98740a40fca613c9f15)

File: lib/iris/coords.py [hunk 3]
```diff
@@ -2194,14 +2194,20 @@ def serialize(x):
...
         ----------
         bound_position : float, default=0.5
             The desired position of the bounds relative to the position
             of the points.
+        monthly : bool, default=False
+            If True, the coordinate must be monthly and bounds are set to the
+            start and ends of each month.
+        yearly : bool, default=False
+            If True, the coordinate must be yearly and bounds are set to the
+            start and ends of each year.
```
Comment (pp-mo) [inline lib/iris/coords.py:2210]: nice !
👍

File: lib/iris/coords.py [hunk 4]
```diff
@@ -2233,31 +2239,78 @@ def _guess_bounds(self, bound_position=0.5):
...
+                        months_and_years.append(date_pair)
+                    else:
+                        raise ValueError(
+                            "Cannot guess monthly bounds for a coordinate with multiple "
+                            "points in a month."
+                        )
+                    lower_bounds.append(date.__class__(lyear, lmonth, 1, 0, 0))
+                    upper_bounds.append(date.__class__(uyear, umonth, 1, 0, 0))
+                bounds = self.units.date2num(np.array([lower_bounds, upper_bounds]).T)
+                contiguous = np.ma.allclose(bounds[1:, 0], bounds[:-1, 1])
```
Comment (pp-mo) [inline lib/iris/coords.py:2269]: I reckon these lines are common to both monthly + yearly, so remove here...
```suggestion
```

File: lib/iris/coords.py [hunk 5]
```diff
@@ -2233,31 +2239,78 @@ def _guess_bounds(self, bound_position=0.5):
...
+                        months_and_years.append(year)
+                    else:
+                        raise ValueError(
+                            "Cannot guess yearly bounds for a coordinate with multiple "
+                            "points in a year."
+                        )
+                    lower_bounds.append(date.__class__(date.year, 1, 1, 0, 0))
+                    upper_bounds.append(date.__class__(date.year + 1, 1, 1, 0, 0))
+                bounds = self.units.date2num(np.array([lower_bounds, upper_bounds]).T)
+                contiguous = np.ma.allclose(bounds[1:, 0], bounds[:-1, 1])
```
Comment (pp-mo) [inline lib/iris/coords.py:2283]: ..and de-indent these ones
```suggestion
            bounds = self.units.date2num(np.array([lower_bounds, upper_bounds]).T)
            contiguous = np.ma.allclose(bounds[1:, 0], bounds[:-1, 1])
```

**Review by pp-mo — CHANGES_REQUESTED** [review]
This is cool, one slight improvement suggested.

But .. I think we need to error if both monthly + yearly are True.
And maybe a mention of that in the docstring too.

File: lib/iris/coords.py (hunk 5 above)
Comment (HGWright) [reply lib/iris/coords.py]: [7ad2cdc](https://github.com/SciTools/iris/pull/6090/commits/7ad2cdc69d3dd63c3fb5391d85a2ef2a0713767c)

File: lib/iris/coords.py (hunk 4 above)
Comment (HGWright) [reply lib/iris/coords.py]: [7ad2cdc](https://github.com/SciTools/iris/pull/6090/commits/7ad2cdc69d3dd63c3fb5391d85a2ef2a0713767c)

**Review by pp-mo — APPROVED** [review]
Nice, I think this nails it all.
Good to have fixed on pytest now, I think -- a mixture of pytest + unittest is not so good !

**Pull request merged**